### PR TITLE
Use SMTP client to send mails

### PR DESCRIPTION
**Pull request by rsto**
This patch rewrites the smtpclient API to use a generic SMTP backend for sending mail.

Up until now, Cyrus forked in the sendmail binary as specified in the `sendmail` imapd.conf option to send email. The new implementation is agnostic of the underlying SMTP backend, using Cyrus protocol streams (lib/prot.h) to abstract the transport.

The new implementation provides TCP-based and sendmail-process based SMTP backends.

Also see: https://github.com/cyrusimap/cyrus-imapd/issues/2182

### CODE REVIEW

**Comment by rsto** [comment]
@ksmurchison May I ask you to review at least the refactored CalDAV, NNTP and Sieve code? I had no chance to test them properly, do you? I might need to set up Cassandane tests for them.

One notable difference is that previously these modules used files to read/compose the email. I have rewritten the code to keep the composed emails in memory until they are sent by the SMTP client. In my understanding the message templates are small, so this shouldn't be a memory issue?

**Comment by rsto** [comment]
> Big question: I wonder if it makes sense for the "host" backend to leverage code that already exists in backend.c ?
> Smaller question: Do we want to standardize the "host" backend on port 587 (SMTP Submission) rather than port 25?

I think the answers to these questions are related. Port 587 -> absolutely! But this brings up the question: do we want to support STARTTLS if the host supports it?

Regarding backend.c: What I like about the current design of smtpclient is that all it reads from and writes to is a lib/prot byte stream. That should make it very simple to plug in any other SMTP backends, and it already came handy with the sendmail and host backends. I don't know enough yet about the backend module if it will allow to keep this level of abstraction.

On the other hand, the STARTTLS support in protocol.h looks promising and will dig deeper in the code to understand I we can reuse it.

Pinging @brong in case he finds time to chime in.

**Comment by ksmurchison** [comment]
backend.c is a generic framework for the client-side of all of the protocols used in Cyrus.  It can parse capabilities and do STARTTLS and authentication.  We already have a profile for LMTP, so I'm sure adding SMTP would be trivial.

File: imap/http_caldav_sched.c [hunk 1]
```diff
@@ -322,23 +321,6 @@ static int imip_send_sendmail(icalcomponent *ical,
         }
     }
 
-    argv[0] = "sendmail";
```
Comment (ajaysusarla) [inline imap/http_caldav_sched.c]: Thanks for remove this piece of code.

**Comment by rsto** [comment]
> backend.c is a generic framework for the client-side of all of the protocols used in Cyrus. It can parse capabilities and do STARTTLS and authentication.

The proposed smtpclient uses prot.h readers/writers to abstract away the underlying backend, be that a forked sendmail process or a TCP host. The backend.h module seems to go the other way round: it sits on top of the prot streams. Would be able to plug in a forked sendmail process under it? It's what most Cyrus installations will use.

This task could actually be a great opportunity for our pair programming session!

**Comment by ksmurchison** [comment]
Yes, we could certainly look into this in a pairing session.

**Comment by rsto** [comment]
@ksmurchison could you please have a look at https://github.com/cyrusimap/cyrus-imapd/pull/2225/commits/e07ac938dfe350b2183b5d54dbbd4ea851f31bb4 ?

Is that all there is to add SASL auth to backend connections? I don't have a setup to test this locally.

**Comment by ksmurchison** [comment]
@rsto That looks about right to me.  That's what mupdate-client.c does.

**Pull request merged**